File: lib/storage/storage.dart
import 'dart:typed_data';
import 'provider.dart';
import 'http_provider.dart';

/// Simple storage interface for Flutter applications.
///
/// Provides [save], [load], [delete], and [exists] operations that route
/// through the configured provider. Default provider is [HttpStorageProvider]
/// for backend proxy.
///
/// Example usage:
/// ```dart
/// final storage = Storage(
///   basePath: 'https://api.example.com/asset/v1/storage',
///   getToken: () async => authService.currentToken,
/// );
///
/// // Save a file
/// await storage.save('images/photo.jpg', imageBytes);
///
/// // Load a file
/// final data = await storage.load('images/photo.jpg');
///
/// // Check if file exists
/// final exists = await storage.exists('images/photo.jpg');
///
/// // Delete a file
/// await storage.delete('images/photo.jpg');
/// ```
class Storage {
  final StorageProvider _provider;

  /// Creates a Storage instance with HTTP backend provider.
  ///
  /// [basePath] - Full URL to storage endpoint (e.g., 'https://api.example.com/asset/v1/storage')
  /// [getToken] - Callback to retrieve the current JWT token
  Storage({
    required String basePath,
    required Future<String> Function() getToken,
  }) : _provider = HttpStorageProvider(
          basePath: basePath,
          getToken: getToken,
        );

  /// Creates a Storage instance with a custom provider.
  ///
  /// Useful for testing or when using a non-standard provider.
  Storage.withProvider(this._provider);

  /// Saves data to the specified path.
  ///
  /// Overwrites any existing file at the path.
  Future<void> save(String path, Uint8List data) => _provider.save(path, data);

  /// Loads data from the specified path.
  ///
  /// Throws [StorageNotFoundException] if the file does not exist.
  Future<Uint8List> load(String path) => _provider.load(path);

  /// Deletes data at the specified path.
  ///
  /// Returns normally if the file does not exist (idempotent).
  Future<void> delete(String path) => _provider.delete(path);

  /// Checks if data exists at the specified path.
  Future<bool> exists(String path) => _provider.exists(path);

  /// Returns the underlying provider.
  StorageProvider get provider => _provider;
}

File: lib/storage/storage.go
package storage

import (
	"strings"

	convCfg "github.com/sofmon/convention/lib/cfg"
	convCtx "github.com/sofmon/convention/lib/ctx"
)

const configKeyBucket convCfg.ConfigKey = "storage_bucket"
const configKeyProvider convCfg.ConfigKey = "storage_provider"
const configKeyCredentials convCfg.ConfigKey = "storage_credentials"

// joinPath combines root and path, handling edge cases with slashes.
// Returns path unchanged if root is empty.
func joinPath(root, path string) string {
	root = strings.Trim(root, "/")
	path = strings.TrimLeft(path, "/")
	if root == "" {
		return path
	}
	return root + "/" + path
}

// Storage provides a simple interface for storing and retrieving files.
// It wraps a Provider and adds context-aware logging and error handling.
type Storage struct {
	provider Provider
	rootPath string // prepended to all paths
}

// New creates a Storage instance using configuration from config files.
// Reads "storage_bucket", "storage_provider" (defaults to "gcs"), and
// "storage_credentials" (JSON key file content) from config.
func New() (s *Storage, err error) {
	bucket, err := convCfg.String(configKeyBucket)
	if err != nil {
		return
	}

	providerName, err := convCfg.String(configKeyProvider)
	if err != nil {
		providerName = "gcs" // default provider
	}

	credentials, err := convCfg.Bytes(configKeyCredentials)
	if err != nil {
		return
	}

	provider, err := NewProvider(providerName, bucket, credentials)
	if err != nil {
		return
	}

	s = &Storage{provider: provider}
	return
}

// NewWithCredentials creates a Storage instance with explicit provider, bucket, and credentials.
// This allows overriding the config-based initialization.
// The credentials parameter should contain the provider-specific authentication data
// (e.g., GCS service account JSON key content).
func NewWithCredentials(ctx convCtx.Context, providerName, bucket string, credentials []byte) (s *Storage, err error) {
	ctx = ctx.WithScope("NewWithCredentials", "provider", providerName, "bucket", bucket)
	defer ctx.Exit(&err)

	provider, err := NewProvider(providerName, bucket, credentials)
	if err != nil {
		return
	}

	s = &Storage{provider: provider}
	return
}

// NewWithProvider creates a Storage instance with a custom provider.
// Useful for testing or when using a non-standard provider.
func NewWithProvider(provider Provider) *Storage {
	return &Storage{provider: provider}
}

// WithRootPath returns a new Storage instance with the specified root path.
// The root path is prepended to all storage paths.
// Example: storage.WithRootPath("tenant-123/data") causes Save(ctx, "file.txt", data)
// to store at "tenant-123/data/file.txt".
func (s *Storage) WithRootPath(rootPath string) *Storage {
	return &Storage{
		provider: s.provider,
		rootPath: joinPath(s.rootPath, rootPath),
	}
}

// RootPath returns the current root path prefix.
func (s *Storage) RootPath() string {
	return s.rootPath
}

// Save stores data at the specified path.
func (s *Storage) Save(ctx convCtx.Context, path string, data []byte) (err error) {
	fullPath := joinPath(s.rootPath, path)
	ctx = ctx.WithScope("storage.Save", "path", fullPath, "size", len(data))
	defer ctx.Exit(&err)

	err = s.provider.Save(ctx, fullPath, data)
	return
}

// Load retrieves data from the specified path.
func (s *Storage) Load(ctx convCtx.Context, path string) (data []byte, err error) {
	fullPath := joinPath(s.rootPath, path)
	ctx = ctx.WithScope("storage.Load", "path", fullPath)
	defer ctx.Exit(&err)

	data, err = s.provider.Load(ctx, fullPath)
	return
}

// Delete removes data at the specified path.
// Returns nil if path does not exist (idempotent).
func (s *Storage) Delete(ctx convCtx.Context, path string) (err error) {
	fullPath := joinPath(s.rootPath, path)
	ctx = ctx.WithScope("storage.Delete", "path", fullPath)
	defer ctx.Exit(&err)

	err = s.provider.Delete(ctx, fullPath)
	return
}

// Exists checks if data exists at the specified path.
func (s *Storage) Exists(ctx convCtx.Context, path string) (exists bool, err error) {
	fullPath := joinPath(s.rootPath, path)
	ctx = ctx.WithScope("storage.Exists", "path", fullPath)
	defer ctx.Exit(&err)

	exists, err = s.provider.Exists(ctx, fullPath)
	return
}

// Provider returns the underlying provider.
func (s *Storage) Provider() Provider {
	return s.provider
}
